### ISSUE TITLE
FasterR-CNN as background and Visibility Score

### ISSUE THREAD

**Issue by SamihaSara** [hasanirtiza/Pedestron#96]
Could you please tell how to set the FasterR-CNN detector and achieve visibility scores of the occluded pedestrians? Thanks.

**Comment by hasanirtiza** [comment]
One way of doing it is that during training you also provide visible bounding box and try to predict the IOU of full-body bounding box and visible-body bounding box (basically your visibility score). If you are wondering that whether you can do it with the current configuration, then no, you have to retrain.

**Comment by SamihaSara** [comment]
My dataset has pedestrians bounding box and corresponding visibility score, what logic should be added to the existing codes to predict the visibility score?
Let me know where in the training codes this could be added.

**Comment by hasanirtiza** [comment]
If you have the corresponding visibility score, then simply add a regression layer and predict visibility score.

**Comment by SamihaSara** [comment]
I appreciate your advice. Could you please tell wherein the current project I should apply the regression layer to predict the visibility score? I am very new to this field, sorry for the silly questions.

**Comment by hasanirtiza** [comment]
You can put it at many places, but say for `Cascade R-CNN`, you can put a layer in [bbox head](https://github.com/hasanirtiza/Pedestron/blob/master/mmdet/models/bbox_heads/bbox_head.py). If you are new, you can see how a classification and regression layer is added on lines 55-57. In your case, it will be a regression layer predicting one number (visibility score) and you can use smoothl1 loss. All you need to do, load the visibility score from dataset and use it as a target for your prediction layer. 

I am closing this issue, if you still cannot do, you might want to create a community request, someone might help.    

P.S no such thing as silly question, there are only silly answers.

**Comment by SamihaSara** [comment]
@hasanirtiza I have a few questions regarding this. Even if after that I cant do this, then will ask for community help. 
I understand I will have to arrange my ground truth according to any of the Caltech,  CityPerson, CrowdHuman,  EuroCity,  or Wider_challenge Format with an added visibility score field.
Does this score have to be inputted simultaneously with the corresponding bounding box during training? 
Can you tell how the classification and regression layer in lines 55-57 are getting class labels and bounding box coordinates directly from annotation files? I am trying to understand how the training dataset is fed into the model.

**Comment by hasanirtiza** [comment]
> Does this score have to be inputted simultaneously with the corresponding bounding box during training?

I would say yes, that should be a straightforward  approach. If your aim is to tell how much a person is occluded/visible than the easiest way would be that use the visibility score along with the bbox. 

> Can you tell how the classification and regression layer in lines 55-57 are getting class labels and bounding box coordinates directly from annotation files? I am trying to understand how the training dataset is fed into the model.

Again have a detailed look at 
[bbox_head.py](https://github.com/hasanirtiza/Pedestron/blob/master/mmdet/models/bbox_heads/bbox_head.py). Have a look inside the functions `def get_target` on line 82-83 and then have a look at the function `loss` at line 97. Additionally, you might want to have a look at detector file as well say for [cascade_rcnn](https://github.com/hasanirtiza/Pedestron/blob/master/mmdet/models/detectors/cascade_rcnn.py) and the dataset file to see [coco.py](https://github.com/hasanirtiza/Pedestron/tree/master/mmdet/datasets) to see how annotations are being fed.
